staging/src/slime.io/slime/modules/meshregistry/pkg/source/zookeeper/watching.go

```diff
@@ -32,228 +34,445 @@ func (s *Source) ServiceNodeDelete(path string) {
+	return item
+}
 
-	s.handleServiceData(s.cache, provider, consumer, service, path)
+func NewServiceWorker(conn *atomic.Value,
+	endpointUpdateFunc func([]string, []string, string),
+	serviceDeleteFunc func(string),
+	gatewatModel bool,
+	initCallbackFactory func() func()) *worker {
+	return &worker{
+		q:     NewQueue(),
+		cache: cmap.New(),
+		epWatcherOpts: EndpointWatcherOpts{
+			conn:                conn,
+			endpointUpdateFunc:  endpointUpdateFunc,
+			serviceDeleteFunc:   serviceDeleteFunc,
+			gatewatModel:        gatewatModel,
+			initCallbackFactory: initCallbackFactory,
+		},
+	}
 }
 
-func labelToSstring(labels map[string]string) (value string) {
-	keys := make([]string, 0)
-	for k := range labels {
-		keys = append(keys, k)
+type worker struct {
+	q     ServiceEventQueue
+	cache cmap.ConcurrentMap
+
+	epWatcherOpts EndpointWatcherOpts
+}
+
+// non block
+func (w *worker) Start(ctx context.Context) {
+	go w.process(ctx)
+}
+
+func (w *worker) HandleEvent(e ServiceEvent) {
+	w.q.Add(e)
+}
+
+func (w *worker) process(ctx context.Context) {
+	for w.processItem(ctx) {
 	}
+}
 
-	sort.Slice(keys, func(i, j int) bool {
-		return keys[i] < keys[j]
-	})
-	var buff bytes.Buffer
-	for _, k := range keys {
-		buff.WriteString(k)
-		buff.WriteString(labels[k])
+func (w *worker) processItem(ctx context.Context) bool {
+	select {
+	case <-ctx.Done():
+		return false
+	default:
+	}
+	e := w.q.Get()
+	watcher, ok := w.cache.Get(e.path)
+	if e.etype == EventTypeDelete && ok {
+		ew := watcher.(*EndpointWatcher)
+		ew.Exit() // may block other service events from being processed？
+		scope.Infof("endpoint watcher %q exited", ew.servicePath)
+		w.cache.Remove(e.path)
+		return true
+	}
+	if e.etype == EventTypeCreate && !ok {
+		ew := NewEndpointWatcher(e.path, w.epWatcherOpts)
+		w.cache.Set(e.path, ew)
+		scope.Infof("service worker start a new endpoint watcher for %s", e.path)
+		ew.Start(ctx)
+		return true
 	}
-	return buff.String()
+	return true
 }
 
-func sortEndpoint(endpoints []*v1alpha3.WorkloadEntry) {
-	sort.Slice(endpoints, func(i, j int) bool {
-		if endpoints[i].Address == endpoints[j].Address {
-			return labelToSstring(endpoints[i].Labels) < labelToSstring(endpoints[j].Labels)
+type ServiceWatcher struct {
+	ctx context.Context
+
+	conn               *atomic.Value
+	rootPath           string
+	endpointUpdateFunc func([]string, []string, string)
+	serviceDeleteFunc  func(string)
+	gatewatModel       bool
+
+	svcs []string
+
+	workers []*worker
+
+	initLock               sync.Mutex
+	initWait               sync.WaitGroup
+	initCnt, initThreshold int
+}
+
+// block until initialized
+func (sw *ServiceWatcher) Start(ctx context.Context) {
+	sw.ctx = ctx
+	initCh := make(chan struct{})
+	go func() {
+		firstLoop := true
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			svcs, _, e, err := sw.conn.Load().(*zk.Conn).ChildrenW(sw.rootPath)
+			if err != nil {
+				// retry and mark ready
```

Comment:
should not mark ready here

Reply:
The `rootPath` may not exist, or it may be some other error. Now we don't exit, but try our best to try again. At this point it is difficult to decide when to mark ready

Reply:
if `err` is not exit we mark ready